lib/seattleflu/id3c/cli/command/etl/redcap_det_swab_n_send.py

```diff
@@ -0,0 +1,575 @@
+import json
+import hashlib
+import requests
+from typing import Any, List, Dict, Optional, Match, Collection
+from datetime import datetime, timezone
+from id3c.db.session import DatabaseSession
+from id3c.db.datatypes import Json
+from id3c.cli.command.etl import etl, race
+
+from id3c.cli.command.etl.redcap_det import (
+    COMPLETE_AND_VERIFIED,
+    LOG,
+    HEADERS,
+    REVISION,
+    mark_skipped,
+    mark_processed,
+    get_redcap_record
+)
+
+REDCAP_URL = 'redcap.iths.org'
+INTERNAL_SYSTEM = "https://seattleflu.org"
+UW_CENSUS_TRACT = '53033005302'
+
+@etl.command("redcap-det-swab-n-send", help = __doc__)
+
+@click.option("--dry-run", "action",
+    help        = "Only go through the motions of changing the database (default)",
+    flag_value  = "rollback",
+    default     = True)
+
+@click.option("--prompt", "action",
+    help        = "Ask if changes to the database should be saved",
+    flag_value  = "prompt")
+
+@click.option("--commit", "action",
+    help        = "Save changes to the database",
+    flag_value  = "commit")
+
+
+def redcap_det_swab_n_send(*, action: str):
+    LOG.debug(f"Starting the REDCap DET ETL routine for Single Swab n Send, revision {REVISION}")
+
+    db = DatabaseSession()
+    PROJECT_ID = '17561'  # TODO use '17421' in production
+    TOKEN_NAME = 'REDCAP_SINGLE_SWAB_AND_SEND_TOKEN'  # TODO use 'REDCAP_SWAB_N_SEND_TOKEN' in production
+    INSTRUMENTS_REQUIRED_TO_BE_COMPLETE = [
+        # 'consent',   # TODO use in production
+        'enrollment_questionnaire',
+        'back_end_mail_scans',
+        # 'illness_questionnaire_nasal_swab_collection',  # TODO use in production
+        'post_collection_data_entry_qc'
+    ]
+
+    # Fetch and iterate over REDCap DET records that aren't processed
+    #
+    # Rows we fetch are locked for update so that two instances of this
+    # command don't try to process the same longitudinal records.
+    LOG.debug("Fetching unprocessed REDCap DET Single Swab n Send records")
+
+    redcap_det = db.cursor("REDCap DET")
+    redcap_det.execute("""
+        select redcap_det_id as id, document
+          from receiving.redcap_det
+         where not processing_log @> %s
+          and document->>'project_id' = %s
+         order by id
+           for update
+        """, (Json([{ "revision": REVISION }]), PROJECT_ID))
+
+    processed_without_error = None
+
+    try:
+        for det in redcap_det:
+            with db.savepoint(f"redcap_det {det.id}"):
+                LOG.info(f"Processing REDCap DET {det.id}")
+
+                redcap_record = get_redcap_record(PROJECT_ID, det.document['record'], TOKEN_NAME)
+
+                for instrument in INSTRUMENTS_REQUIRED_TO_BE_COMPLETE:
+                    if redcap_record[f"{instrument}_complete"] != COMPLETE_AND_VERIFIED:
+                        LOG.debug(f"Skipping incomplete or unverified REDCap DET {det.id}")
+                        mark_skipped(db, det.id)
+                        continue
```

Comment:
This `continue` is within a nested for loop, so the process doesn't move on to the next `det`

Reply:
Good catch